### PR TITLE
Set Default values correctly for docker variables

### PR DESCRIPTION
**Pull request by denraf**
##### SUMMARY
Environment variables were no longer respected as the default is set twice.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
docker common utils

##### ANSIBLE VERSION
ansible 2.6.0
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Feb 20 2018, 09:19:12) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]

##### ADDITIONAL INFORMATION
env:
export DOCKER_TLS_VERIFY=1
export DOCKER_HOST=tcp://192.168.99.20:2376
export DOCKER_CERT_PATH=~/.docker/certs
export DOCKER_DEFAULT_IP=192.168.99.20
export DOCKER_TLS_HOSTNAME="docker-dev"

      docker_container:
        name: "{{ item.name }}"
        docker_host: "{{ item.docker_host | default('unix://var/run/docker.sock') }}"
        hostname: "{{ item.name }}"
        image: "molecule_local/{{ item.image }}"
        state: started
        recreate: false
        log_driver: json-file
        command: "{{ item.command | default('bash -c \"while true; do sleep 10000; done\"') }}"
        privileged: "{{ item.privileged | default(omit) }}"
        security_opts: "{{ item.security_opts | default(omit) }}"
        volumes: "{{ item.volumes | default(omit) }}"
        tmpfs: "{{ item.tmpfs | default(omit) }}"
        capabilities: "{{ item.capabilities | default(omit) }}"
        exposed_ports: "{{ item.exposed_ports | default(omit) }}"
        published_ports: "{{ item.published_ports | default(omit) }}"
        ulimits: "{{ item.ulimits | default(omit) }}"
        networks: "{{ item.networks | default(omit) }}"
        dns_servers: "{{ item.dns_servers | default(omit) }}"
      register: server
      with_items: "{{ molecule_yml.platforms }}"
      async: 7200
      poll: 0

Before:
"msg": "Error connecting: Error while fetching server API version: ('Connection aborted.', BadStatusLine('\\x15\\x03\\x01\\x00\\x02\\x02\\n',))"

After:
changed: [localhost] => (item={'_ansible_parsed': True, '_ansible_item_result': True, '_ansible_item_label': {'exposed_ports': [u'22/tcp', u'22/udp', u'80/tcp', u'443/tcp'], 'dns_servers': [u'10.86.2.1'], 'name': u'haproxy1', 'image': u'dev-redhat-base:latest', 'hostname': u'haproxy1', 'capabilities': [u'NET_ADMIN'], 'docker_host': u'tcp://192.168.99.20:2376', 'command': u'/sbin/init', 'registry': {'url': u'192.168.99.12:6559'}, 'groups': [u'haproxy'], 'volumes': [u'/sys/fs/cgroup:/sys/fs/cgroup:ro'], 'tmpfs': [u'/tmp', u'/run']}, u'ansible_job_id': u'525370450142.2904', 'failed': False, u'started': 1, 'changed': True, 'item': {'exposed_ports': [u'22/tcp', u'22/udp', u'80/tcp', u'443/tcp'], 'dns_servers': [u'192.168.99.1'], 'name': u'haproxy1', 'image': u'dev-redhat-base:latest', 'hostname': u'haproxy1', 'capabilities': [u'NET_ADMIN'], 'docker_host': u'tcp://192.168.99.20:2376', 'command': u'/sbin/init', 'registry': {'url': u'192.168.99.12:6559'}, 'groups': [u'haproxy'], 'volumes': [u'/sys/fs/cgroup:/sys/fs/cgroup:ro'], 'tmpfs': [u'/tmp', u'/run']}, u'finished': 0, u'results_file': u'/root/.ansible_async/525370450142.2904', '_ansible_ignore_errors': None, '_ansible_no_log': False}) => {    }
    changed: [localhost] => {
        "censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result",
        "changed": true
    }
    META: ran handlers

### CODE REVIEW

**Comment by felixfontein** [comment]
Thanks for pointing this problem out! There is a much better solution, though:
```
from ansible.module_utils.basic import env_fallback

[...]
DOCKER_COMMON_ARGS = dict(
    docker_host=dict(type='str', aliases=['docker_url'], default=DEFAULT_DOCKER_HOST, fallback=(env_fallback, 'DOCKER_TLS_HOSTNAME')),
[...]
```
This will take care of extracting the value from the environment variable if it is defined and the parameter is not specified.

(To see more examples, grep for `fallback=` in `lib/ansible/modules`.)

File: lib/ansible/module_utils/docker_common.py
Comment (felixfontein) [inline lib/ansible/module_utils/docker_common.py]: This needs a fallback with `DOCKER_API_VERSION`

File: lib/ansible/module_utils/docker_common.py
Comment (felixfontein) [inline lib/ansible/module_utils/docker_common.py]: The fallback env variable is called `DOCKER_TLS`, not `DOCKER_TLS_VERIFY` (that's the one below).

**Review by felixfontein — CHANGES_REQUESTED** [review]
Almost done, I think :)

**Comment by felixfontein** [comment]
@denraf You need to remove the merge commits. [Check the link for instructions](https://docs.ansible.com/ansible/devel/dev_guide/developing_rebasing.html).

File: lib/ansible/module_utils/docker_common.py
Comment (felixfontein) [inline lib/ansible/module_utils/docker_common.py]: Ah, sorry, I forgot one more: `timeout` is missing the `DOCKER_TIMEOUT` fallback.

**Review by felixfontein — CHANGES_REQUESTED** [review]

**Comment by felixfontein** [comment]
`ssl_version=dict(type='str', fallback=(env_fallback, 'DOCKER_SSL_VERSION')),` and `ssl_version=self._get_value('ssl_version', params['ssl_version'], 'DOCKER_SSL_VERSION', None),` (as before) should do the trick.

**Comment by felixfontein** [comment]
There are some tests failing which look strange. The original task executed is:
```
- name: Parameter name should be required
  docker_secret:
    state: present
  ignore_errors: yes
  register: output
```
The verbose task result is:
```
fatal: [testhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "api_version": "/tmp/ansible-test-coverage-9VVQrx/coverage/injector.py", 
            "docker_host": "/tmp/ansible-test-coverage-9VVQrx/coverage/injector.py", 
            "state": "present", 
            "timeout": "/tmp/ansible-test-coverage-9VVQrx/coverage/injector.py", 
            "tls": "/tmp/ansible-test-coverage-9VVQrx/coverage/injector.py", 
            "tls_hostname": "/tmp/ansible-test-coverage-9VVQrx/coverage/injector.py", 
            "tls_verify": "/tmp/ansible-test-coverage-9VVQrx/coverage/injector.py"
        }
    }, 
    "msg": "parameters are mutually exclusive: tls, tls_verify"
}
```
I wonder whether this is caused by the tests, or if it happens every time a `docker_` module is used.

**Comment by felixfontein** [comment]
I can reproduce this with `docker_container` from this branch:
```
fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "api_version": "/usr/bin/python",
            "docker_host": "/usr/bin/python",
            "image": "felixfontein/acme-test-container:25448686e9b499e42380ddf965d8e23bd794378c",
            "kernel_memory": 0,
            "memory": "512M",
            "memory_reservation": "4M",
            "memory_swap": "1G",
            "name": "ansible-test-container",
            "state": "started",
            "timeout": "/usr/bin/python",
            "tls": "/usr/bin/python",
            "tls_hostname": "/usr/bin/python",
            "tls_verify": "/usr/bin/python"
        }
    },
    "msg": "parameters are mutually exclusive: tls, tls_verify"
}
```

**Comment by felixfontein** [comment]
Ooookay, this seems to be a bug in the environment fallback system.

**Comment by felixfontein** [comment]
Ooops. Nope, even more stupid: I copied the example wrongly... The environment variable's name must be in a list. Now, it iterates over the env var name (and thus over its letters) and checks for environment variables called `D`, `O`, `C`, `K`, `E`, `R`, `_`, ..., and as soon as it finds one, it takes the contents of that one...

**Comment by felixfontein** [comment]
It must look like this:
```
DOCKER_COMMON_ARGS = dict(
    docker_host=dict(type='str', aliases=['docker_url'], default=DEFAULT_DOCKER_HOST, fallback=(env_fallback, ['DOCKER_HOST'])),
    tls_hostname=dict(type='str', default=DEFAULT_TLS_HOSTNAME, fallback=(env_fallback, ['DOCKER_TLS_HOSTNAME'])),
    api_version=dict(type='str', aliases=['docker_api_version'], default='auto', fallback=(env_fallback, ['DOCKER_API_VERSION'])),
    timeout=dict(type='int', default=DEFAULT_TIMEOUT_SECONDS, fallback=(env_fallback, ['DOCKER_TIMEOUT'])),
    cacert_path=dict(type='str', aliases=['tls_ca_cert']),
    cert_path=dict(type='str', aliases=['tls_client_cert']),
    key_path=dict(type='str', aliases=['tls_client_key']),
    ssl_version=dict(type='str'),
    tls=dict(type='bool', default=DEFAULT_TLS, fallback=(env_fallback, ['DOCKER_TLS'])),
    tls_verify=dict(type='bool', default=DEFAULT_TLS_VERIFY, fallback=(env_fallback, ['DOCKER_TLS_VERIFY'])),
    debug=dict(type='bool', default=False)
)
```

**Comment by denraf** [comment]
Thank you for your help @felixfontein !

**Comment by felixfontein** [comment]
@denraf And thanks a lot for noticing and fixing this! :)

**Review by felixfontein — APPROVED** [review]
shipit

**Comment by felixfontein** [comment]
@kassiansun Can you take a look here?

**Comment by kassiansun** [comment]
shipit

**Comment by felixfontein** [comment]
bot_status

**Comment by kassiansun** [comment]
@felixfontein Seems we're not maintainers of `docker_common`

**Comment by felixfontein** [comment]
Yep. Well, I guess we need help then :)

@resmo Can you take a look?

**Comment by resmo** [comment]
Hi all, is this a regression? Made by 4117b2dd297b170d7b81bdcf613ae71049a4139e ?

**Comment by resmo** [comment]
AFAICS, there are more things to "fix" or cleanup. IMHO the source of the issue is in `_get_value()`
https://github.com/ansible/ansible/pull/42641/files#diff-1ba35f3163ddb36a0cfc16c50579434dR204

But for a "hotfix" this PR looks good to me. We should merge this and cleanup in 2.7 only.

shipit

**Comment by felixfontein** [comment]
@resmo Ah, you found the source ;) Yes, it looks like it.

I think the best way is to remove the env variable parameter from `_get_value()`, and use the fallback mechanism for all variables. But that's something for 2.7 or maybe later (since Core Freeze is today).

**Comment by felixfontein** [comment]
bot_status

File: lib/ansible/module_utils/docker_common.py
Comment (gundalow) [inline lib/ansible/module_utils/docker_common.py:86]: Could you please update `lib/ansible/utils/module_docs_fragments/docker.py` to mention these variables, see `lib/ansible/utils/module_docs_fragments/vmware.py` to an example

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/docker_common.py
Comment (felixfontein) [reply lib/ansible/module_utils/docker_common.py:86]: Yes, that would be a good idea. I guess we can get rid of the corresponding paragraph in the `notes:` section then.

@denraf Can you also update the documentation (in the docs fragment) for `cacert_path`, `cert_path` and `key_path`? They are a bit more tricky, since there the value of the env variable `DOCKER_CERT_PATH` is taken and modified.

**Comment by gundalow** [comment]
After discussion with @felixfontein I'll merge this and @felixfontein will update this in a follow up PR

**Review by gundalow — APPROVED** [review]

**Pull request merged**
